object_store/src/gcp/mod.rs

```diff
@@ -591,6 +592,52 @@ impl ObjectStore for GoogleCloudStorage {
         Ok((upload_id, Box::new(CloudMultiPartUpload::new(inner, 8))))
     }
 
+    async fn start_multipart(&self, location: &Path) -> Result<MultipartId> {
+        self.client.multipart_initiate(location).await
+    }
+
+    /// Upload a part of a multi-part upload.
+    /// Caller is responsible for the validity of the upload_id and part_number.
+    async fn add_multipart(
+        &self,
+        location: &Path,
+        upload_id: &MultipartId,
+        part_number: usize,
+        bytes: Bytes,
+    ) -> Result<UploadPart> {
+        let encoded_path =
+            percent_encode(location.to_string().as_bytes(), NON_ALPHANUMERIC).to_string();
+        let inner = GCSMultipartUpload {
+            client: Arc::clone(&self.client),
+            encoded_path,
+            multipart_id: upload_id.clone(),
+        };
+        // TODO: figure out the right way to handle error types.
+        // The multipart upload returns an IO error, but this wants a crate-specific error.
+        Ok(inner
+            .put_multipart_part(bytes.to_vec(), part_number)
+            .await
+            .unwrap())
```

Comment:
I don't think we want to unwrap here (or anywhere in this code) on errors as that will panic our code (which expects, at least in the controller, to be able to handle errors).

Generally you would handle this by wrapping the IO error in the crate error type. For example, there's a Error::Generic type that can handle any kind of error.

Reply:
Ah, figured it out. The `Error` in the gcp context is GCP specific and I was having a hard time constructing it. The crate's Error does the trick.